neqo-transport/src/ecn.rs

```diff
@@ -0,0 +1,213 @@
+    baseline: EcnCount,
+}
+
+impl EcnInfo {
+    /// Set the baseline (= the ECN counts from the last ACK Frame).
+    pub fn set_baseline(&mut self, baseline: EcnCount) {
+        self.baseline = baseline;
+    }
+
+    /// Expose the current baseline.
+    pub fn baseline(&self) -> EcnCount {
+        self.baseline
+    }
+
+    /// Count the number of packets sent out on this path during ECN validation.
+    /// Exit ECN validation if the number of packets sent exceeds `ECN_TEST_COUNT`.
+    /// We do not implement the part of the RFC that says to exit ECN validation if the time since
+    /// the start of ECN validation exceeds 3 * PTO, since this seems to happen much too quickly.
+    pub fn on_packet_sent(&mut self) {
+        if self.state != EcnValidationState::Testing {
+            return;
+        }
+
+        self.probes_sent += 1;
+        if self.probes_sent == ECN_TEST_COUNT {
+            qdebug!(
+                "ECN probing concluded with {} packet sent",
+                self.probes_sent
+            );
+            self.state = EcnValidationState::Unknown;
+        }
+    }
+
+    /// After the ECN validation test has ended, check if the path is ECN capable.
+    pub fn validate_ack_ecn(&mut self, acked_packets: &[SentPacket], ack_ecn: Option<EcnCount>) {
+        // RFC 9000, Appendix A.4:
+        //
+        // > From the "unknown" state, successful validation of the ECN counts in an ACK frame
+        // > (see Section 13.4.2.1) causes the ECN state for the path to become "capable", unless
+        // > no marked packet has been acknowledged.
+        if self.state != EcnValidationState::Unknown {
+            return;
+        }
+
+        // RFC 9000, Section 13.4.2.1:
+        //
+        // > Validating ECN counts from reordered ACK frames can result in failure. An endpoint MUST
+        // > NOT fail ECN validation as a result of processing an ACK frame that does not increase
+        // > the largest acknowledged packet number.
+        let largest_acked = acked_packets.first().expect("must be there").pn;
+        if largest_acked <= self.largest_acked {
+            return;
+        }
+
+        // RFC 9000, Section 13.4.2.1:
+        //
+        // > An endpoint that receives an ACK frame with ECN counts therefore validates
+        // > the counts before using them. It performs this validation by comparing newly
+        // > received counts against those from the last successfully processed ACK frame.
+        //
+        // > If an ACK frame newly acknowledges a packet that the endpoint sent with
+        // > either the ECT(0) or ECT(1) codepoint set, ECN validation fails if the
+        // > corresponding ECN counts are not present in the ACK frame.
+        if ack_ecn.is_none() {
+            qwarn!("ECN validation failed, no ECN counts in ACK frame");
+            self.state = EcnValidationState::Failed;
+            return;
+        }
+        let ack_ecn = ack_ecn.unwrap();
```

Comment:
Might be less error prone without the unwrap:

```suggestion
        let Some(ack_ecn) = ack_ecn else {
            qwarn!("ECN validation failed, no ECN counts in ACK frame");
            self.state = EcnValidationState::Failed;
            return;
        };
```